modules/server/src/main/scala/higherkindness/compendium/core/CompendiumService.scala

```diff
@@ -25,23 +25,28 @@ import higherkindness.compendium.storage.Storage
 trait CompendiumService[F[_]] {
 
   def storeProtocol(id: String, protocol: Protocol): F[Unit]
-  def recoverProtocol(id: String): F[Option[Protocol]]
+  def recoverProtocol(protocolId: String): F[Option[Protocol]]
+  def existsProtocol(protocolId: String): F[Boolean]
 }
 
 object CompendiumService {
 
-  def impl[F[_]: Sync: Storage: DBService: ProtocolUtils] = new CompendiumService[F] {
+  implicit def impl[F[_]: Sync: Storage: DBService: ProtocolUtils](): CompendiumService[F] =
+    new CompendiumService[F] {
 
-    val utils = ProtocolUtils[F]
+      override def storeProtocol(id: String, protocol: Protocol): F[Unit] =
+        ProtocolUtils[F].validateProtocol(protocol) >>
+          DBService[F].upsertProtocol(id, protocol) >>
+          Storage[F].store(id, protocol)
 
-    override def storeProtocol(id: String, protocol: Protocol): F[Unit] =
-      ProtocolUtils[F].validateProtocol(protocol) >>
-        DBService[F].addProtocol(id, protocol) >>
-        Storage[F].store(id, protocol)
+      override def recoverProtocol(protocolId: String): F[Option[Protocol]] =
+        DBService[F]
+          .existsProtocol(protocolId)
```

Comment:
Isn't this quite elaborate?
```scala
override def existsProtocol(id: String): F[Boolean] = Storage[F].exists(id)
```

Where

```scala
 override def exists(id: String): F[Boolean] =
        Sync[F].catchNonFatal(new File(s"${config.path}${File.separator}$id")).map(_.exists)
```

But if we check `Storage[F].recover` we are doing this:
```scala
for {
          filename <- Sync[F].catchNonFatal {
            Option(new File(s"${config.path}${File.separator}$id").listFiles())
              .fold(Option.empty[String])(_.headOption.map(_.getAbsolutePath))
          }
        ...
```

So, if the file does not exist, then recover will return an `F[None]`. Aren't we doing this method redundant? 🤔

Reply:
No at all. Bear in mind, that `FileDBService` is an example of the DBService functionality. Ideally, we need to create [another](https://github.com/higherkindness/compendium/pull/46) typeclass using `postgresql` instead of `file`.

The same goes for storage. In the future we will have multiple typeclasses for each of the possible storages: SchemaRegistry, postgresql, ...

If we want to know if a protocol exists, we ask the dbservice. It would not make any sense to make a query to the storage.

Reply:
Oh, I see I see... So then there's no problem! Thanks! 😄